### PR TITLE
[ENH, DOC] Add Euclidean distance with convolution for similarity search

### PR DESCRIPTION
**Pull request by baraline**
#### What does this implement/fix? Explain your changes.

- Add computation of Euclidean and squared Euclidean distance with dot product computed by convolution. 
- Add tests to check the consistency of the results between naive and speedup version of distance profile computations
- Fix speed-up argument usage for BaseSimilaritySearch class, add a function to return all available speed-ups.
- Improve docstring and variable names consistency overall in the module
- Add the option to apply the exclusion zone to the prediction of the best matches. For example, in the case of TopKSimilaritySearch, it means that when k>1, the returned best matches will be spaced by at least `query_length//exclusion_factor` timestamps when selected from the same sample. This avoids cases where the best matches are direct neighbours.
- Update the associated notebooks, compare the performance of naive and convolution version, show how to use the new `apply_exclusion_to_result` parameter of predict.

#### Does your contribution introduce a new dependency? If yes, which one?

No additional dependency.

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [X] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).

##### For developers with write access
- [X] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/similarity_search/distance_profiles/_commons.py [hunk 1]
```diff
@@ -43,20 +43,20 @@ def fft_sliding_dot_product(X, q):
...
     -------
-    output : shape=(n_features, n_timestamps - (length - 1))
+    output : shape=(n_features, series_length - length + 1)
         Sliding dot product between q and X.
     """
-    n_features, n_timestamps = X.shape[0]
+    n_features, n_timestamps = X.shape
     length = q.shape[1]
-    out = np.zeros((n_features, n_timestamps - (length - 1)))
+    out = np.zeros((n_features, n_timestamps - length + 1))
```
Comment (TonyBagnall) [inline aeon/similarity_search/distance_profiles/_commons.py:59]: seem to be mixing series_length and n_timestamps. In the other modules we use n_timepoints (and n_channels if multivariate)

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
all looks good, but we should consider n_timepoints vs series_length. I think I would prefer package wide standardisation, but I dont really mind series_length. There is I think still one instance of n_timestamps though :)

File: aeon/similarity_search/distance_profiles/_commons.py (hunk 1 above)
Comment (baraline) [reply aeon/similarity_search/distance_profiles/_commons.py:59]: Thanks ! I thought the standard in aeon was series_length and not n_timepoints, am I mistaking ?
Seems I focused on docstring more than variable names, will make some standardisation

Comment (baraline) [reply aeon/similarity_search/distance_profiles/_commons.py]: Should be all fixed now !

**Review by TonyBagnall — APPROVED** [review]
LGTM

**Pull request merged**